File: backend/functions/callback.js
/**
 * Callback Lambda Function
 * Handles the OAuth callback and stores credentials in Parameter Store
 * Supports multiple app registrations via dynamic credential loading
 */

const crypto = require('crypto');
const { getAppCredentials, storeShopCredentials } = require('./credentials');

/**
 * Parse state parameter to extract app identifier
 * @param {string} state - Base64 encoded JSON state
 * @returns {{nonce: string, app: string}}
 */
function parseState(state) {
  try {
    const decoded = Buffer.from(state, 'base64').toString('utf8');
    return JSON.parse(decoded);
  } catch (error) {
    throw new Error('Invalid state parameter');
  }
}

/**
 * Exchange authorization code for access token
 */
async function exchangeCodeForToken(shop, code, clientId, clientSecret) {
  const response = await fetch(`https://${shop}/admin/oauth/access_token`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      client_id: clientId,
      client_secret: clientSecret,
      code: code,
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Failed to exchange code: ${response.status} ${response.statusText} - ${errorText}`);
  }

  return await response.json();
}

/**
 * Verify HMAC signature from Shopify
 */
function verifyHmac(queryParams, hmac, apiSecret) {
  // Remove hmac and signature from params
  const params = { ...queryParams };
  delete params.hmac;
  delete params.signature;

  // Sort and build query string
  const sortedParams = Object.keys(params)
    .sort()
    .map(key => `${key}=${params[key]}`)
    .join('&');

  // Generate HMAC
  const hash = crypto
    .createHmac('sha256', apiSecret)
    .update(sortedParams)
    .digest('hex');

  return hash === hmac;
}

exports.handler = async (event) => {
  try {
    const params = event.queryStringParameters || {};
    const { code, hmac, shop, state } = params;

    // Validate required parameters
    if (!code || !hmac || !shop || !state) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'Missing required parameters' }),
      };
    }

    // Parse state to get app identifier
    let stateData;
    try {
      stateData = parseState(state);
    } catch (error) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'Invalid state parameter' }),
      };
    }

    const { app } = stateData;
    if (!app) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'Missing app identifier in state' }),
      };
    }

    // Load app credentials from Parameter Store
    let appCredentials;
    try {
      appCredentials = await getAppCredentials(app);
    } catch (error) {
      console.error(`Failed to load credentials for app ${app}:`, error.message);
      return {
        statusCode: 404,
        body: JSON.stringify({ error: 'App not found', message: `No credentials found for app: ${app}` }),
      };
    }

    // Verify HMAC with the correct app secret
    if (!verifyHmac(params, hmac, appCredentials.clientSecret)) {
      console.error(`HMAC verification failed for shop ${shop}, app ${app}`);
      return {
        statusCode: 403,
        body: JSON.stringify({ error: 'Invalid HMAC signature' }),
      };
    }

    // Exchange code for access token
    console.log(`Exchanging code for token: shop=${shop}, clientId=${appCredentials.clientId}`);
    const tokenData = await exchangeCodeForToken(
      shop,
      code,
      appCredentials.clientId,
      appCredentials.clientSecret
    );

    console.log(`Token exchange response: access_token prefix=${tokenData.access_token?.substring(0, 10)}, scope=${tokenData.scope}`);

    // Store credentials in Parameter Store under /shops/{appId}/{shop}/
    await storeShopCredentials(app, shop, tokenData.access_token, tokenData.scope);

    console.log(`Successfully stored credentials for shop: ${shop}, app: ${app}`);

    // Redirect to frontend with success message
    const frontendUrl = process.env.FRONTEND_URL;
    return {
      statusCode: 302,
      headers: {
        Location: `${frontendUrl}?shop=${shop}&app=${app}&installed=true`,
      },
      body: '',
    };
  } catch (error) {
    console.error('Callback error:', error);
    return {
      statusCode: 500,
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        error: 'Failed to complete installation',
        message: error.message
      }),
    };
  }
};

File: backend/functions/callback.py
"""
Callback Lambda Function
Handles the OAuth callback and stores credentials in Parameter Store
"""

import os
import json
import hmac
import hashlib
from datetime import datetime

import boto3
import requests

ssm_client = boto3.client('ssm', region_name=os.environ.get('AWS_REGION', 'eu-central-1'))


def exchange_code_for_token(shop: str, code: str) -> dict:
    """Exchange authorization code for access token"""
    api_key = os.environ.get('SHOPIFY_API_KEY')
    api_secret = os.environ.get('SHOPIFY_API_SECRET')

    response = requests.post(
        f'https://{shop}/admin/oauth/access_token',
        json={
            'client_id': api_key,
            'client_secret': api_secret,
            'code': code
        },
        headers={'Content-Type': 'application/json'}
    )

    if not response.ok:
        raise Exception(f'Failed to exchange code: {response.text}')

    return response.json()


def store_credentials(shop: str, access_token: str, scopes: str):
    """Store credentials in AWS Parameter Store"""
    prefix = os.environ.get('PARAMETER_STORE_PREFIX', '/shopify/clients')

    # Store access token
    ssm_client.put_parameter(
        Name=f'{prefix}/{shop}/access-token',
        Value=access_token,
        Type='SecureString',
        Description=f'Shopify access token for {shop}',
        Overwrite=True,
        Tags=[
            {'Key': 'App', 'Value': 'Shopify'},
            {'Key': 'Shop', 'Value': shop}
        ]
    )

    # Store scopes for reference
    ssm_client.put_parameter(
        Name=f'{prefix}/{shop}/scopes',
        Value=scopes,
        Type='String',
        Description=f'Shopify scopes for {shop}',
        Overwrite=True,
        Tags=[
            {'Key': 'App', 'Value': 'Shopify'},
            {'Key': 'Shop', 'Value': shop}
        ]
    )

    # Store installation timestamp
    ssm_client.put_parameter(
        Name=f'{prefix}/{shop}/installed-at',
        Value=datetime.utcnow().isoformat(),
        Type='String',
        Description=f'Installation timestamp for {shop}',
        Overwrite=True,
        Tags=[
            {'Key': 'App', 'Value': 'Shopify'},
            {'Key': 'Shop', 'Value': shop}
        ]
    )


def verify_hmac(query_params: dict, hmac_value: str) -> bool:
    """Verify HMAC signature from Shopify"""
    api_secret = os.environ.get('SHOPIFY_API_SECRET')

    # Remove hmac and signature from params
    params = {k: v for k, v in query_params.items() if k not in ('hmac', 'signature')}

    # Sort and build query string
    sorted_params = '&'.join(f'{k}={v}' for k, v in sorted(params.items()))

    # Generate HMAC
    computed_hmac = hmac.new(
        api_secret.encode('utf-8'),
        sorted_params.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(computed_hmac, hmac_value)


def handler(event, context):
    try:
        params = event.get('queryStringParameters') or {}
        code = params.get('code')
        hmac_value = params.get('hmac')
        shop = params.get('shop')

        # Validate required parameters
        if not code or not hmac_value or not shop:
            return {
                'statusCode': 400,
                'body': json.dumps({'error': 'Missing required parameters'})
            }

        # Verify HMAC
        if not verify_hmac(params, hmac_value):
            return {
                'statusCode': 403,
                'body': json.dumps({'error': 'Invalid HMAC signature'})
            }

        # Exchange code for access token
        token_data = exchange_code_for_token(shop, code)

        # Store credentials in Parameter Store
        store_credentials(shop, token_data['access_token'], token_data['scope'])

        print(f'Successfully stored credentials for shop: {shop}')

        # Redirect to frontend with success message
        frontend_url = os.environ.get('FRONTEND_URL', '')
        return {
            'statusCode': 302,
            'headers': {
                'Location': f'{frontend_url}?shop={shop}&installed=true'
            },
            'body': ''
        }
    except Exception as e:
        print(f'Callback error: {e}')
        return {
            'statusCode': 500,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({
                'error': 'Failed to complete installation',
                'message': str(e)
            })
        }
